matscipy/calculators/manybody/calculator.py

```diff
@@ -83,122 +109,135 @@ def __init__(self, atom_type, pair_type, F, G, d1F, d2F, d11F, d22F, d12F, d1G,
+            raise TypeError(
+                f"{self.neighbourhood} is not of type Neighbourhood")
+
     def get_cutoff(self, atoms):
+        """Return a valid cutoff."""
         if np.isscalar(self.cutoff):
             return self.cutoff
 
-        # get internal atom types from atomic numbers
-        elements = set(atoms.numbers)
-
-        # loop over all possible element combinations
-        cutoff = 0
-        for i in elements:
-            ii = self.atom_type(i)
-            for j in elements:
-                jj = self.atom_type(j)
-                p = self.pair_type(ii, jj)
-                cutoff = max(cutoff, self.cutoff[p])
-        return cutoff
+        # Get largest cutoff for all possible pair types
+        atom_types = map(self.atom_type, np.unique(atoms.numbers))
+        pair_types = starmap(self.pair_type, product(atom_types, repeat=2))
+        return max(self.cutoff[p] for p in pair_types)
 
     def calculate(self, atoms, properties, system_changes):
-        Calculator.calculate(self, atoms, properties, system_changes)
+        """Calculate system properties."""
+        super().calculate(atoms, properties, system_changes)
+
+        # Setting up cutoffs for neighbourhood
+        self.neighbourhood.cutoff = self.get_cutoff(atoms)  # no-op on mols
 
         # get internal atom types from atomic numbers
-        t_n = self.atom_type(atoms.numbers)
-        cutoff = self.get_cutoff(atoms)
+        t_n = self.neighbourhood.atom_type(atoms.numbers)
 
         # construct neighbor list
-        i_p, j_p, r_p, r_pc = neighbour_list('ijdD', atoms=atoms, cutoff=cutoff)
+        i_p, j_p, r_p, r_pc = self.neighbourhood.get_pairs(atoms, "ijdD")
 
         nb_atoms = len(self.atoms)
         nb_pairs = len(i_p)
 
         # normal vectors
-        n_pc = (r_pc.T / r_p).T
+        n_pc = r_pc / r_p[_c]
 
         # construct triplet list
-        first_n = first_neighbours(nb_atoms, i_p)
-        ij_t, ik_t = triplet_list(first_n)
+        ij_t, ik_t, r_tq, r_tqc = self.neighbourhood.get_triplets(atoms,
+                                                                  "ijdD")
 
         # construct lists with atom and pair types
         ti_p = t_n[i_p]
-        tij_p = self.pair_type(ti_p, t_n[j_p])
+        tij_p = self.neighbourhood.pair_type(ti_p, t_n[j_p])
         ti_t = t_n[i_p[ij_t]]
-        tij_t = self.pair_type(ti_t, t_n[j_p[ij_t]])
-        tik_t = self.pair_type(ti_t, t_n[j_p[ik_t]])
+        tij_t = self.neighbourhood.pair_type(ti_t, t_n[j_p[ij_t]])
+        tik_t = self.neighbourhood.pair_type(ti_t, t_n[j_p[ik_t]])
 
         # potential-dependent functions
-        G_t = self.G(r_pc[ij_t], r_pc[ik_t], ti_t, tij_t, tik_t)
-        d1G_tc = self.d1G(r_pc[ij_t], r_pc[ik_t], ti_t, tij_t, tik_t)
-        d2G_tc = self.d2G(r_pc[ij_t], r_pc[ik_t], ti_t, tij_t, tik_t)
+        G_t = self.G(r_tqc[:, 0], r_tqc[:, 1], ti_t, tij_t, tik_t)
+        d1G_tc = self.d1G(r_tqc[:, 0], r_tqc[:, 1], ti_t, tij_t, tik_t)
```

Comment:
I don't seem to understand what a nuplet is (see my comment above), hence it is difficult to interpret what 0 and 1 mean here. I guess this refers to the argument of the `G`-function?

Reply:
Ah, it's the index in the triplet, i.e. whether you are referring to the first or second bond in the triplet.

Reply:
I guess this will run from 0..2 once we switch to absolute distances rather than vectors.

Reply:
Yes exactly. I'll change `nuplet` to `n-uplet`, it's just a generic way of calling triplets, quadruplets, etc.